lib/edge/src/query.rs

```diff
@@ -110,30 +122,45 @@ impl Shard {
                 }
 
                 Source::Prefetch(merge_plan) => {
-                    let merged = self
-                        .recurse_prefetch(
-                            *merge_plan,
-                            search_results,
-                            scroll_results,
-                            depth + 1,
-                            hw_counter_acc.clone(),
-                        )?
-                        .into_iter();
-
-                    sources.extend(merged);
+                    let merged = self.recurse_prefetch(
+                        *merge_plan,
+                        search_results,
+                        scroll_results,
+                        depth + 1,
+                        hw_counter_acc.clone(),
+                    )?;
+
+                    sources.push(merged);
                 }
             }
         }
 
-        // Rescore or return plain sources
-        if let Some(rescore_params) = rescore_params {
-            let rescored = self.rescore(sources, rescore_params, hw_counter_acc)?;
-            Ok(vec![rescored])
+        if let Some(rescore_stages) = rescore_stages {
+            let RescoreStages {
+                shard_level,
+                collection_level,
+            } = rescore_stages;
+
+            let shard_stage_result = if let Some(rescore_params) = shard_level {
+                vec![self.rescore(sources, rescore_params, hw_counter_acc.clone())?]
+            } else {
+                sources
+            };
+
+            let collection_result = if let Some(rescore_params) = collection_level {
+                self.rescore(shard_stage_result, rescore_params, hw_counter_acc)?
+            } else {
+                shard_stage_result.into_iter().next().unwrap_or_default()
```

Comment:
Do we only expect length 0 or 1? Then I'd prefer some sort of conversion that enforces it. **Or** a comment, that we only use first result, even when there are more than one available.